shared/pubsub.js

```diff
@@ -426,15 +468,30 @@ const publicMethods = {
     }
     // Remove WebSocket event listeners.
     if (this.socket) {
-      for (const name of eventNames) {
+      for (const name of socketEventNames) {
         this.socket.removeEventListener(name, this.listeners[name])
       }
       this.socket.close()
     }
+    this.listeners = {}
     this.socket = null
     this.shouldReconnect = false
   },
 
+  // Emits a custom event or an `error` event.
+  // Other fake native events are not allowed so as to not break things.
+  emit (type: string, detail?: mixed) {
+    if (!customEventNames.includes(type) && type !== 'error') {
+      throw new Error(`emit(): argument 'type' must not be '${type}'.`)
+    }
+    // This event object partially implements the `CustomEvent` interface.
+    const event = { type, detail }
+    const listener = this.listeners[type]
+    if (listener) {
+      listener(event)
+    }
+  },
+
```

Comment:
I can't tell for sure, but this code seems like it only supports one listener per event?

Is there a specific reason you chose to re-implement this different type of event system, or do you think you can use the existing `okTurtes.events` domain instead (as is used throughout the app)? I feel like that would be cleaner and avoid re-inventing the wheel. `okTurtles.events` also handles multiple listeners per event.

Reply:
Yes there is a single listener per event, but that listener can actually wrap several handler functions.

However, to keep the API simple for now it only allows one default "hard-coded" handler and one optional custom (e.g. "user-defined") handler though, whereas traditional API's allow to dynamically add and remove an unlimited number of custom handlers.

Reply:
So is there an issue with using `okTurtles.events`?

Reply:
Say we want to use `okTurtles.events` to make the general banner be cleared and dissappear upon successful reconnection.

For now this can be done like this, from anywhere in the frontend code:

```js
sbp('okTurtles.events/on', 'reconnection-succeeded', (event) => {
  sbp('okTurtles.data/get', 'BANNER').clean()
})

sbp('okTurtles.data/apply', PUBSUB_INSTANCE, (client) => {
  client.customEventHandlers['reconnection-succeeded'] = (event) => {
    sbp('okTurtles.events/emit', 'reconnection-succeeded', event)
  }
})
```

The `reconnection-succeeded` event type name can also be made a constant for consistency.

Reply:
After thinking a bit more about it, it looks like a tighter sbp events integration could also be achieved by deleting my `emit()` method and using `okTurtles.events/emit()` instead.

Reply:
> After thinking a bit more about it, it looks like a tighter sbp events integration could also be achieved by deleting my emit() method and using okTurtles.events/emit() instead.

Go for it 👍

Reply:
While it is easy to implement in the browser, it becomes more complicated in the backend unit tests, where there are commonly several pubsub client instances running at once and pubsub events are not global but are only scoped to their respective client instance.

Therefore the instance must be passed somehow when emitting pubsub events, so what would be the best way to do that in SBP?
- Passing the instance as an argument. This requires to define and remember which argument is used, e.g. it could the first positional argument, or its `target` property (assuming the argument is an object).
- Passing an instance identifier in the selector, e.g. `okTurtles.events/pubsub-5d498f/emit`

Reply:
@snowteamer I'm not fully sure I understand the issue, maybe you can link me (either here or on Slack) to the part in the tests where this is happening?

FYI the implementation of `emit` allows for the passing of various variables to the listeners, so you could even pass in `this` and have the listener take in parameters. Copying the implementation of `emit` here for reference:

```
  'okTurtles.events/emit': function (event: string, ...data: any) {
    for (const listener of sbp('okTurtles.data/get', listenKey(event)) || []) {
      listener(...data)
    }
  },
```

Not sure if that helps you or not.